### ISSUE TITLE
Fatal exception editing note category

### ISSUE THREAD

**Issue by dan0xii** [stefan-niedermann/nextcloud-notes#303]
Open a note, open the context menu and then select "Category". Attempt to type anything and you may see the first character appear and then the app crashes. When a note already has a category set that category will be displayed correctly until you type something and then you get the crash. I've only experienced this on emulators as I don't have access to try and replicate on my phone. I've seen this on API 19/4.4 and API 25/7.1.1 with my live instance of nc. Logcats below:

4.4
```
12-04 22:27:03.279 3112-3112/it.niedermann.owncloud.notes D/dalvikvm: VFY: replacing opcode 0x6f at 0x0008
12-04 22:27:03.279 3112-3112/it.niedermann.owncloud.notes I/dalvikvm: Could not find method android.widget.Button.setAutoSizeTextTypeUniformWithConfiguration, referenced from method android.support.v7.widget.AppCompatButton.setAutoSizeTextTypeUniformWithConfiguration
12-04 22:27:03.279 3112-3112/it.niedermann.owncloud.notes W/dalvikvm: VFY: unable to resolve virtual method 21014: Landroid/widget/Button;.setAutoSizeTextTypeUniformWithConfiguration (IIII)V
12-04 22:27:03.279 3112-3112/it.niedermann.owncloud.notes D/dalvikvm: VFY: replacing opcode 0x6f at 0x0006
12-04 22:27:03.279 3112-3112/it.niedermann.owncloud.notes I/dalvikvm: Could not find method android.widget.Button.setAutoSizeTextTypeUniformWithPresetSizes, referenced from method android.support.v7.widget.AppCompatButton.setAutoSizeTextTypeUniformWithPresetSizes
12-04 22:27:03.279 3112-3112/it.niedermann.owncloud.notes W/dalvikvm: VFY: unable to resolve virtual method 21015: Landroid/widget/Button;.setAutoSizeTextTypeUniformWithPresetSizes ([II)V
12-04 22:27:03.279 3112-3112/it.niedermann.owncloud.notes D/dalvikvm: VFY: replacing opcode 0x6f at 0x0006
12-04 22:27:03.279 3112-3112/it.niedermann.owncloud.notes I/dalvikvm: Could not find method android.widget.Button.setAutoSizeTextTypeWithDefaults, referenced from method android.support.v7.widget.AppCompatButton.setAutoSizeTextTypeWithDefaults
12-04 22:27:03.279 3112-3112/it.niedermann.owncloud.notes W/dalvikvm: VFY: unable to resolve virtual method 21016: Landroid/widget/Button;.setAutoSizeTextTypeWithDefaults (I)V
12-04 22:27:03.279 3112-3112/it.niedermann.owncloud.notes D/dalvikvm: VFY: replacing opcode 0x6f at 0x0006
12-04 22:27:03.589 3112-3112/it.niedermann.owncloud.notes D/EGL_emulation: eglMakeCurrent: 0xb8458b70: ver 3 1
12-04 22:27:03.589 3112-3112/it.niedermann.owncloud.notes E/EGL_emulation: tid 3112: eglSurfaceAttrib(1199): error 0x3009 (EGL_BAD_MATCH)
12-04 22:27:03.589 3112-3112/it.niedermann.owncloud.notes W/HardwareRenderer: Backbuffer cannot be preserved
12-04 22:27:03.629 3112-3112/it.niedermann.owncloud.notes D/EGL_emulation: eglMakeCurrent: 0xb8458b70: ver 3 1
12-04 22:27:03.689 3112-3112/it.niedermann.owncloud.notes D/EGL_emulation: eglMakeCurrent: 0xb8458b70: ver 3 1
12-04 22:27:03.989 3112-3112/it.niedermann.owncloud.notes D/EGL_emulation: eglMakeCurrent: 0xb8458b70: ver 3 1
12-04 22:27:04.539 3112-3112/it.niedermann.owncloud.notes D/EGL_emulation: eglMakeCurrent: 0xb8458b70: ver 3 1
12-04 22:27:05.639 3112-3143/it.niedermann.owncloud.notes W/Filter: An exception occured during performFiltering()!
                                                                    java.lang.NullPointerException
                                                                        at it.niedermann.owncloud.notes.android.fragment.CategoryDialogFragment$FolderArrayAdapter$FolderFilter.performFiltering(CategoryDialogFragment.java:153)
                                                                        at android.widget.Filter$RequestHandler.handleMessage(Filter.java:234)
                                                                        at android.os.Handler.dispatchMessage(Handler.java:102)
                                                                        at android.os.Looper.loop(Looper.java:136)
                                                                        at android.os.HandlerThread.run(HandlerThread.java:61)
12-04 22:27:05.729 3112-3112/it.niedermann.owncloud.notes D/AndroidRuntime: Shutting down VM
12-04 22:27:05.729 3112-3112/it.niedermann.owncloud.notes W/dalvikvm: threadid=1: thread exiting with uncaught exception (group=0x9ccd0b20)
12-04 22:27:05.729 3112-3112/it.niedermann.owncloud.notes E/AndroidRuntime: FATAL EXCEPTION: main
                                                                            Process: it.niedermann.owncloud.notes, PID: 3112
                                                                            java.lang.NullPointerException
                                                                                at java.util.ArrayList.addAll(ArrayList.java:188)
                                                                                at android.widget.ArrayAdapter.addAll(ArrayAdapter.java:195)
                                                                                at it.niedermann.owncloud.notes.android.fragment.CategoryDialogFragment$FolderArrayAdapter$FolderFilter.publishResults(CategoryDialogFragment.java:184)
                                                                                at android.widget.Filter$ResultsHandler.handleMessage(Filter.java:282)
                                                                                at android.os.Handler.dispatchMessage(Handler.java:102)
                                                                                at android.os.Looper.loop(Looper.java:136)
                                                                                at android.app.ActivityThread.main(ActivityThread.java:5017)
                                                                                at java.lang.reflect.Method.invokeNative(Native Method)
                                                                                at java.lang.reflect.Method.invoke(Method.java:515)
                                                                                at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:779)
                                                                                at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:595)
                                                                                at dalvik.system.NativeStart.main(Native Method)

```

7.1.1

```
12-04 22:43:35.264 4489-4511/it.niedermann.owncloud.notes D/EGL_emulation: eglMakeCurrent: 0x792fa7ef56e0: ver 3 1 (tinfo 0x792fa7eb4bc0)
12-04 22:43:35.288 4489-4511/it.niedermann.owncloud.notes E/eglCodecCommon: glUtilsParamSize: unknow param 0x000082da
12-04 22:43:35.288 4489-4511/it.niedermann.owncloud.notes E/eglCodecCommon: glUtilsParamSize: unknow param 0x00008cdf
12-04 22:43:35.288 4489-4511/it.niedermann.owncloud.notes E/eglCodecCommon: glUtilsParamSize: unknow param 0x00008824
12-04 22:43:35.416 4489-4511/it.niedermann.owncloud.notes D/EGL_emulation: eglMakeCurrent: 0x792fa7ef56e0: ver 3 1 (tinfo 0x792fa7eb4bc0)
12-04 22:43:50.804 4489-4494/it.niedermann.owncloud.notes I/art: Do partial code cache collection, code=20KB, data=30KB
12-04 22:43:50.804 4489-4494/it.niedermann.owncloud.notes I/art: After code cache collection, code=20KB, data=30KB
12-04 22:43:50.804 4489-4494/it.niedermann.owncloud.notes I/art: Increasing code cache capacity to 128KB
12-04 22:43:50.945 4489-4489/it.niedermann.owncloud.notes I/com.yydcdut.rxmarkdown.RxMDEditText: finish-->7
12-04 22:43:52.382 4489-4489/it.niedermann.owncloud.notes I/Choreographer: Skipped 83 frames!  The application may be doing too much work on its main thread.
12-04 22:43:52.395 4489-4511/it.niedermann.owncloud.notes D/EGL_emulation: eglMakeCurrent: 0x792fa7ef56e0: ver 3 1 (tinfo 0x792fa7eb4bc0)
12-04 22:43:54.714 4489-4511/it.niedermann.owncloud.notes D/EGL_emulation: eglMakeCurrent: 0x792fa7ef56e0: ver 3 1 (tinfo 0x792fa7eb4bc0)
12-04 22:43:54.902 4489-4511/it.niedermann.owncloud.notes D/EGL_emulation: eglMakeCurrent: 0x792fa7ef56e0: ver 3 1 (tinfo 0x792fa7eb4bc0)
12-04 22:43:54.935 4489-4511/it.niedermann.owncloud.notes D/EGL_emulation: eglMakeCurrent: 0x792fa7ef56e0: ver 3 1 (tinfo 0x792fa7eb4bc0)
12-04 22:43:54.999 4489-4511/it.niedermann.owncloud.notes D/EGL_emulation: eglMakeCurrent: 0x792fa7ef56e0: ver 3 1 (tinfo 0x792fa7eb4bc0)
12-04 22:43:55.040 4489-4489/it.niedermann.owncloud.notes W/IInputConnectionWrapper: finishComposingText on inactive InputConnection
12-04 22:43:55.176 4489-4511/it.niedermann.owncloud.notes D/EGL_emulation: eglMakeCurrent: 0x792fa7ef56e0: ver 3 1 (tinfo 0x792fa7eb4bc0)
12-04 22:43:56.545 4489-4511/it.niedermann.owncloud.notes D/EGL_emulation: eglMakeCurrent: 0x792fa7ef56e0: ver 3 1 (tinfo 0x792fa7eb4bc0)
12-04 22:43:56.676 4489-4511/it.niedermann.owncloud.notes D/EGL_emulation: eglMakeCurrent: 0x792fa7ef56e0: ver 3 1 (tinfo 0x792fa7eb4bc0)
12-04 22:43:59.828 4489-4511/it.niedermann.owncloud.notes D/EGL_emulation: eglMakeCurrent: 0x792fa7ef56e0: ver 3 1 (tinfo 0x792fa7eb4bc0)
12-04 22:43:59.861 4489-4511/it.niedermann.owncloud.notes D/EGL_emulation: eglMakeCurrent: 0x792fa7ef56e0: ver 3 1 (tinfo 0x792fa7eb4bc0)
12-04 22:44:01.616 4489-4520/it.niedermann.owncloud.notes W/Filter: An exception occured during performFiltering()!
                                                                    java.lang.NullPointerException: Attempt to invoke interface method 'int java.util.List.size()' on a null object reference
                                                                        at it.niedermann.owncloud.notes.android.fragment.CategoryDialogFragment$FolderArrayAdapter$FolderFilter.performFiltering(CategoryDialogFragment.java:153)
                                                                        at android.widget.Filter$RequestHandler.handleMessage(Filter.java:234)
                                                                        at android.os.Handler.dispatchMessage(Handler.java:102)
                                                                        at android.os.Looper.loop(Looper.java:154)
                                                                        at android.os.HandlerThread.run(HandlerThread.java:61)
12-04 22:44:01.635 4489-4489/it.niedermann.owncloud.notes D/AndroidRuntime: Shutting down VM
12-04 22:44:01.635 4489-4489/it.niedermann.owncloud.notes E/AndroidRuntime: FATAL EXCEPTION: main
                                                                            Process: it.niedermann.owncloud.notes, PID: 4489
                                                                            java.lang.NullPointerException: Attempt to invoke interface method 'java.lang.Object[] java.util.Collection.toArray()' on a null object reference
                                                                                at java.util.ArrayList.addAll(ArrayList.java:562)
                                                                                at android.widget.ArrayAdapter.addAll(ArrayAdapter.java:222)
                                                                                at it.niedermann.owncloud.notes.android.fragment.CategoryDialogFragment$FolderArrayAdapter$FolderFilter.publishResults(CategoryDialogFragment.java:184)
                                                                                at android.widget.Filter$ResultsHandler.handleMessage(Filter.java:282)
                                                                                at android.os.Handler.dispatchMessage(Handler.java:102)
                                                                                at android.os.Looper.loop(Looper.java:154)
                                                                                at android.app.ActivityThread.main(ActivityThread.java:6119)
                                                                                at java.lang.reflect.Method.invoke(Native Method)
                                                                                at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:886)
                                                                                at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:776)
```

So that I'm not breaking any etiquette:
```
**Android version:** 4.4/7.1.1, see above
**Device**: only in the emulator
**System language**: English (US)
**App version:** master branch
**App source:** self-built
**Steps to reproduce:**
See above
```

**Comment by stefan-niedermann** [comment]
Strange, i can not reproduce this on my smartphone (7.1.1), but with the stack trace i should be able to find the problem.

**Comment by dan0xii** [comment]
I couldn't at first but it seems to have started doing it on my phone as well. I've found that I can reproduce it by doing:

1. I start the app and open a note (I can't tell if there's a difference between which note is used) 
2. Try to edit the category and the drop-down list of available categories will be shown. 
3. Cancel that dialog and then close the note
4. Re-open the note and try and do the same thing again and the drop-down won't be shown. As soon as you try to type the category text it'll crash.

**Comment by dan0xii** [comment]
I've been looking into this further and this is what I've found. It seems to happen more frequently on a low spec emulator than it does on my phone which is fairly fast. The NPE comes from publishResults in CategoryDialogFragment when "results" is null. I think this is caused by originalData being null so when performFiltering is called it ends up returning null, possibly because the asynctask hasn't finished loading originalData.

**Comment by korelstar** [comment]
I was able to reproduce this after adding a 5 seconds sleep in the LoadCategoriesTask. Fix was easy, see #323 😄
